### PR TITLE
[Unity] Scaffolding DLight

### PR DESCRIPTION
**Pull request by junrushao**
This PR introduces a new package `tvm.dlight`. DLight provides infra to implement composable default schedules to TVM IRModule

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+        while True:
+            try:
+                func = py_mod.Module[f"func{i}"]
+            except:  # pylint: disable=bare-except
+                break
+            else:
+                print(f"Working on {py_mod}::func{i}")
+            i += 1
+            mod = IRModule.from_expr(func.with_attr("global_symbol", "main"))
+            schedules = gen.generate_design_space(mod)
```
Comment (tqchen) [inline main.py:259]: Let us structure the testcase UT same as Before/After test, e.g. one softmax function before and after with structural equality test.

Atm we only test that the schedule exists

File: python/tvm/dlight/testing/modules.py [hunk 2]
```diff
@@ -0,0 +1,2621 @@
+from tvm.script import ir as I
+from tvm.script import tir as T
+
+# fmt: off
+
+@I.ir_module
+class Decode:
+
```
Comment (tqchen) [inline python/tvm/dlight/testing/modules.py:8]: let us not have testing here, instead move to `tests/python/dlight/` so we can grow test cases organically

File: python/tvm/dlight/gpu/default_schedule.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+def _default_schedule(  # pylint: disable=too-many-locals
+    func: tir.PrimFunc,
+    target: Target,
+) -> tir.PrimFunc:
+    max_threads_per_block = None
+    for name in ["max_threads_per_block", "max_num_threads"]:
+        if max_threads_per_block is None:
+            max_threads_per_block = target.attrs.get(name, None)
+    if max_threads_per_block is None:
```
Comment (tqchen) [inline python/tvm/dlight/gpu/default_schedule.py:19]: Should have one schedule_rule, and invoke that schedule rule here. The schedule do not need to match all instances. Dlight is meant to get performant schedules

File: python/tvm/dlight/gpu/default_schedule.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        max_threads_per_block = 64
+
+    get_loop_iter_type = get_global_func("tir.schedule.GetLoopIterType")
```
Comment (tqchen) [inline python/tvm/dlight/gpu/default_schedule.py:22]: This can be a python function get exposed

File: python/tvm/dlight/testing/modules.py (hunk 2 above)
Comment (junrushao) [reply python/tvm/dlight/testing/modules.py:8]: got it

File: python/tvm/dlight/gpu/default_schedule.py (hunk 3 above)
Comment (junrushao) [reply python/tvm/dlight/gpu/default_schedule.py:19]: that makes sense

File: python/tvm/dlight/testing/modules.py (hunk 2 above)
Comment (junrushao) [reply python/tvm/dlight/testing/modules.py:8]: moved to `tests/python/dlight/test_modules.py`

File: python/tvm/dlight/gpu/default_schedule.py (hunk 3 above)
Comment (junrushao) [reply python/tvm/dlight/gpu/default_schedule.py:19]: i'm going to remove this rule first. seems not super relevant to this PR

File: tests/python/dlight/test_modules.py [hunk 5]
```diff
@@ -0,0 +1,2621 @@
...
+                var_compute_intermediate_1[v_i0, v_i1, v_i2] = var_compute_intermediate[v_i0, v_i1, v_i2]
+        for ax0, ax1, ax2 in T.grid(T.int64(1), T.int64(1), T.int64(2560)):
+            with T.block("T_add_1"):
+                v_ax0, v_ax1, v_ax2 = T.axis.remap("SSS", [ax0, ax1, ax2])
+                T.reads(var_compute_intermediate_1[v_ax0, v_ax1, v_ax2], lv701[v_ax0, v_ax1, v_ax2])
+                T.writes(var_T_add_intermediate_1[v_ax0, v_ax1, v_ax2])
+                var_T_add_intermediate_1[v_ax0, v_ax1, v_ax2] = var_compute_intermediate_1[v_ax0, v_ax1, v_ax2] + lv701[v_ax0, v_ax1, v_ax2]
+        for i0, i1, i2 in T.grid(T.int64(1), T.int64(1), T.int64(2560)):
+            with T.block("compute_2"):
+                v_i0, v_i1, v_i2 = T.axis.remap("SSS", [i0, i1, i2])
```
Comment (tqchen) [inline tests/python/dlight/test_modules.py:369]: Would be good to delete this for now, and as we add new test cases like `test_normalize.py` we add these modules as before/after in each test file.

Comment (junrushao) [reply tests/python/dlight/test_modules.py]: will do

**Pull request merged**